Update OptionalELResolver to support method invocation

pnicolucci opened the pull request:
fixes #27122

pnicolucci committed:
Update OptionalELResolver to support method invocation

pnicolucci committed:
Add OptionalELResolver FAT

pnicolucci commented:
#libby #build

volosied commented on dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java:
```diff
@@ -0,0 +1,571 @@
+    }
+
+    /**
+     * Add the OptionalELResolver to the list of ELResolver.
+     *
+     * Evaluate an Optional property that is not Empty.
```
```suggestion
     * Add the OptionalELResolver to the list of ELResolver.
     *
     * Evaluate an Optional property that is not Empty.
     *
     *  Scenario: "If the Optional.isPresent() is true for the base object and the property is not null then the resulting value is the result of calling ELResolver.getValue(ELContext, Object, Object) with the following parameters:

    The ELContext is the current context
    The base object is the result of calling Optional.get() on the current base object
    The property object is the current property object"
    * 
```

volosied replied:
Asterisks are missing -- text is long and the formatting it in the suggestion is difficult.

volosied replied:
You're right -- this scenario doesn't apply here. It applies to `testEL60OptionalELResolver_OptionalBeanProperty_WithOptionalELResolver` instead.

volosied commented on dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java:
```diff
@@ -0,0 +1,571 @@
+     * Add the OptionalELResolver to the list of ELResolver.
+     *
+     * Evaluate an Optional String property that is Empty.
+     *
```
```suggestion
     * Add the OptionalELResolver to the list of ELResolver.
     *
     * Evaluate an Optional String property that is Empty.
     *
     * Scenario: "If the Optional.isEmpty() is true for the base object and the property is null then the resulting value is null."
     *
```

volosied commented on dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java:
```diff
@@ -0,0 +1,571 @@
+     * Do not add the OptionalELResolver to the list of ELResolver.
+     *
+     * Evaluate an Optional String property that is Empty.
+     *
```
```suggestion
     * Do not add the OptionalELResolver to the list of ELResolver.
     *
     * Evaluate an Optional String property that is Empty.
     *
     * Scenario: "If the Optional.isEmpty() is true for the base object and the property is not null then the resulting value is the base object (an empty Optional). " 
     *
```

volosied reviewed (commented):
Tests look great.  I added a few small suggestions which state the scenarios in the Javadoc. 

The only thing that's puzzling me is  this scenario:

"If the Optional.isPresent() is true for the base object and the property is null then the resulting value is the result of calling Optional.get() on the base object. " 

I'm confusing myself over this. Does a test cover this?  Value is there, but property is null. If a property is not defined on bean, is it still considered null? 


edit: 

Line for reference: https://github.com/apache/tomcat/blob/eeb91237cb35dc1c9e9b5e51476df57f694bc44c/java/jakarta/el/OptionalELResolver.java#L65 

I still can't figure out what this scenario covers, but I think the current tests are adequate enough.

volosied commented:
It's not necessary to add the javadoc to each, but I though the first few were useful for easy reference (at least when I was looking over the tests).

pnicolucci committed:
Update comments for OptionalELResolver FAT

Add a new test case to verify PropertyNotFoundException when using
an Optional bean Property.

volosied reviewed (approved):
Look good to me. Thanks for the additional javadoc!

pnicolucci commented:
#libby

pnicolucci commented:
#build

pnicolucci commented:
#build

pnicolucci commented:
#libby #build
